File: Habitica/src/main/java/com/habitrpg/android/habitica/data/local/implementation/RealmTagLocalRepository.java
package com.habitrpg.android.habitica.data.local.implementation;

import com.habitrpg.android.habitica.data.local.TagLocalRepository;
import com.habitrpg.android.habitica.models.Tag;

import java.util.List;

import io.realm.OrderedRealmCollectionSnapshot;
import io.realm.Realm;
import io.realm.RealmResults;
import rx.Observable;


public class RealmTagLocalRepository extends RealmBaseLocalRepository implements TagLocalRepository {
    public RealmTagLocalRepository(Realm realm) {
        super(realm);
    }

    @Override
    public Observable<RealmResults<Tag>> getTags(String userId) {
        return getRealm().where(Tag.class).equalTo("userId", userId).findAll().asObservable();
    }

    @Override
    public void removeOldTags(List<Tag> onlineTags, String userID) {
        OrderedRealmCollectionSnapshot<Tag> localTags = getRealm().where(Tag.class).equalTo("userId", userID).findAll().createSnapshot();
        for (Tag localTag : localTags) {
            if (!onlineTags.contains(localTag)) {
                localTag.deleteFromRealm();
            }
        }
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/data/local/implementation/RealmTagLocalRepository.kt
package com.habitrpg.android.habitica.data.local.implementation

import com.habitrpg.android.habitica.data.local.TagLocalRepository
import com.habitrpg.android.habitica.models.Tag
import io.realm.Realm
import kotlinx.coroutines.flow.Flow

class RealmTagLocalRepository(
    realm: Realm,
) : RealmBaseLocalRepository(realm),
    TagLocalRepository {
    override fun deleteTag(tagID: String) {
        val tags = safeQuery { it.where(Tag::class.java).equalTo("id", tagID) }?.findAll() ?: return
        executeTransaction { tags.deleteAllFromRealm() }
    }

    override fun getTags(userId: String): Flow<List<Tag>> = safeFindAll {
        it.where(Tag::class.java).equalTo("userId", userId)
    }
}
